### PR TITLE
Enable RBAC checking on controllers

### PR DESCRIPTION
**Pull request by hsong-rh**
This is the sixth (last) PR to enable RBAC support in Approval service:

1. Add resource level RBAC checking;
2. Add three endpoints: `/admin/requests`, `/approver/requests` and `/owner/requests` for different roles to list requests;

### CODE REVIEW

File: app/controllers/api/v1x0/mixins/rbac_mixin.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        end
+
+        def resource_check(verb, id = params[:id], klass = controller_name.classify.constantize)
+          access_obj = permission_check(verb, klass)
+
+          raise Exceptions::NotAuthorizedError, "#{verb.titleize} access not authorized for #{klass}" if !access_obj.admin? && access_obj.not_owned?(id.to_i) && access_obj.not_approvable?(id.to_i)
+        end
+
+        def permission_check(verb, klass = controller_name.classify.constantize)
+          return unless RBAC::Access.enabled?
```
Comment (bzwei) [inline app/controllers/api/v1x0/mixins/rbac_mixin.rb]: what shall you return if not `enabled?`? will it cause exception in line 24?

File: app/controllers/api/v1x0/templates_controller.rb [hunk 2]
```diff
@@ -2,6 +2,9 @@ module Api
   module V1x0
     class TemplatesController < ApplicationController
       include Mixins::IndexMixin
+      include Mixins::RBACMixin
+
+      before_action :read_access_check, :only => %i[show index]
```
Comment (bzwei) [inline app/controllers/api/v1x0/templates_controller.rb]: you already include `index` in `before_action`, do you still need #152 for templates_controller?

Comment (hsong-rh) [reply app/controllers/api/v1x0/templates_controller.rb]: Will remove `index`

File: app/controllers/api/v1x0/mixins/rbac_mixin.rb (hunk 1 above)
Comment (hsong-rh) [reply app/controllers/api/v1x0/mixins/rbac_mixin.rb]: `resource_check` should also have `return unless RBAC::Access.enabled?` in the beginning.

File: app/controllers/api/v1x0/mixins/rbac_mixin.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
+module Api
+  module V1x0
+    module Mixins
+      module RBACMixin
+        def create_access_check
```
Comment (bzwei) [inline app/controllers/api/v1x0/mixins/rbac_mixin.rb]: Please add comments why some use `permission_check` while others use `resource_check`. This is a mixin therefore it should be more generic.
Create, destroy, and update likely need the same permission, while read may require a different one.

File: spec/controllers/v1.0/workflows_controller_spec.rb [hunk 4]
```diff
@@ -27,8 +41,18 @@
       end
     end
 
-    context 'when template does not exist' do
+    context 'admin role when template does not exist' do
       let!(:template_id) { 0 }
+      let(:access_obj) { instance_double(RBAC::Access, :accessible? => true, :admin? => true, :approver? => false, :owner? => false) }
```
Comment (mkanoor) [inline spec/controllers/v1.0/workflows_controller_spec.rb]: 👍 the instance double really helps here

File: spec/controllers/v1.0/templates_controller_spec.rb [hunk 5]
```diff
@@ -11,26 +11,53 @@
...
+    context 'when admin role' do
+      let(:access_obj) { instance_double(RBAC::Access, :accessible? => true, :admin? => true, :approver? => false, :owner? => false) }
+
+      before do
+        allow(access_obj).to receive(:not_owned?).and_return(false)
+        allow(access_obj).to receive(:not_approvable?).and_return(false)
+        allow(access_obj).to receive(:approver_id_list).and_return([])
+        allow(access_obj).to receive(:owner_id_list).and_return([])
+
+        get "#{api_version}/templates", :params => { :limit => 5, :offset => 0 }, :headers => request_header
```
Comment (mkanoor) [inline spec/controllers/v1.0/templates_controller_spec.rb]: @hsong-rh Should the default headers be in a single spec/support file like here
https://github.com/ManageIQ/catalog-api/blob/cee85e285bec837b21cc0fe6723097bdfbd2539c/spec/support/service_spec_helper.rb#L7

File: spec/controllers/v1.0/workflows_controller_spec.rb [hunk 6]
```diff
@@ -41,23 +65,46 @@
...
   describe "GET /workflows with filter" do
+    let(:access_obj) { instance_double(RBAC::Access, :accessible? => true, :admin? => true, :approver? => false, :owner? => false) }
     before do
+      allow(RBAC::Access).to receive(:new).with('workflows', 'read').and_return(access_obj)
+      allow(access_obj).to receive(:process).and_return(access_obj)
+      allow(access_obj).to receive(:not_owned?).and_return(false)
+      allow(access_obj).to receive(:not_approvable?).and_return(false)
+      allow(access_obj).to receive(:approver_id_list).and_return([])
+      allow(access_obj).to receive(:owner_id_list).and_return([])
+
```
Comment (mkanoor) [inline spec/controllers/v1.0/workflows_controller_spec.rb]: @hsong-rh Are the limit and offset required parameters in the line below, couldn't we just use the defaults

File: spec/controllers/v1.0/requests_controller_spec.rb [hunk 7]
```diff
@@ -159,24 +274,99 @@
...
   # Test suite for POST /workflows/:workflow_id/requests
   describe 'POST /workflows/:workflow_id/requests' do
     let(:item) { { 'disk' => '100GB' } }
-    let(:valid_attributes) { { :requester_name => '1234', :name => 'Visit Narnia', :content => item, :description => 'desc' } }
+    let(:valid_attributes) { { :owner => '1234', :name => 'Visit Narnia', :content => item } }
+
+    context 'admin role when request attributes are valid' do
+      let(:access_obj) { instance_double(RBAC::Access, :accessible? => true, :admin? => true, :approver? => false, :owner? => false) }
+      before do
+        ENV['AUTO_APPROVAL'] = 'y'
```
Comment (mkanoor) [inline spec/controllers/v1.0/requests_controller_spec.rb]: @hsong-rh we could use with_modified_env from climate control gem here

**Comment by bzwei** [comment]
Follow-up PRs will be created to add back resolved spec tests.

**Pull request merged**